src/Keboola/DbExtractor/Retry/SqlRetryPolicy.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\DbExtractor\Retry;
+
+use Retry\Policy\AbstractRetryPolicy;
+use Retry\RetryContextInterface;
+
+class SqlRetryPolicy extends AbstractRetryPolicy
+{
+    /**
+     * The default limit to the number of attempts for a new policy.
+     *
+     * @var int
+     */
+    public const DEFAULT_MAX_ATTEMPTS = 3;
+
+    /**
+     * The maximum number of retry attempts before failure.
+     *
+     * @var int
+     */
+    private $maxAttempts;
+
+    /**
+     * The list of retryable exceptions
+     *
+     * @var array
+     */
+    private $retryableExceptions = ['Exception'];
+
+    /**
+     * The list of SQLSTATE code masks to NOT retry
+     *
+     * @var array
+     */
+    private $ignorableSqlstateMasks = [];
+
+    /**
+     * @param int        $maxAttempts The number of attempts before a retry becomes impossible.
+     * @param array|null $retryableExceptions
+     */
+    public function __construct(
+        ?int $maxAttempts = null,
+        ?array $retryableExceptions = null,
+        ?array $ignorableSqlstateMasks = null
+    ) {
+
+        $this->maxAttempts = $maxAttempts ?? self::DEFAULT_MAX_ATTEMPTS;
+
+        if ($retryableExceptions) {
+            $this->retryableExceptions = $retryableExceptions;
+        }
+
+        if ($ignorableSqlstateMasks) {
+            $this->ignorableSqlstateMasks = $ignorableSqlstateMasks;
+        }
+    }
+
+    public function canRetry(RetryContextInterface $context): bool
+    {
+        $e = $context->getLastException();
+
+        return (!$e || $this->shouldRetryForException($e)) && $context->getRetryCount() < $this->maxAttempts;
+    }
+
+    private function shouldRetryForException(\Exception $e): bool
+    {
+        foreach ($this->retryableExceptions as $class) {
+            if (is_a($e, $class)) {
+                if ($this->shouldIgnoreForSqlStateCode($e->getMessage())) {
+                    return false;
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private function shouldIgnoreForSqlStateCode(string $errorMessage): bool
+    {
+        if (strstr($errorMessage, 'SQLSTATE') === false) {
+            return false;
+        }
+        preg_match('/SQLSTATE\[(.*)\]/', $errorMessage, $matches);
+        if (count($matches) < 2) {
+            return false;
+        }
+        $sqlCode = $matches[1];
+        foreach ($this->ignorableSqlstateMasks as $ignorable) {
+            preg_match('/' . $ignorable . '/', $sqlCode, $ignorableMatches);
```

Comment:
afaik, all `^42` level codes should be not-retryable because they're defined as user errors https://en.wikipedia.org/wiki/SQLSTATE.  So, I think it could be useful to just put 42, instead of writing out each individual one.  I was thinking of using something like `42***` but then thought why not just use regex.

Reply:
Also, I think it's easier to list the codes that we *don't* want to retry, and have the default be to retry.

Reply:
ok, so from the other perspective - i see only 08000 to 08007 as retryable

also checked this https://docs.oracle.com/cd/A97630_01/appdev.920/a58231/appd.htm
which might possibly add 61000

Reply:
and of course 0 would be retryable, becauese that's what mysql returns consistently :)

Reply:
good question, i don't know.

What about a little research - run oracle,mysql,pgsql ex with 
- invalid password, invalid server, non-existent table, syntax error in query?

thought i'm not sure it will really help us